introduce callForSeenUsers and countSeenUsers

butonic opened the pull request:
I was running into the upgrade trying to iterate over 170000 zombie users in my two test ldap servers, when only 6 of them ever logged in. Instead of waiting 5-6 hours or maybe even longer I decided to take the time and find a better solution.

This pr introduces callForSeenUsers() and countSeenUsers() to allow only executing a callback for users that logged in at least once and as a result have a home directory that has been initialized. It also helps prevent out of memory errors.

cc @PVince81 @DeepDiver1975 there are at least two further occurences (SyncService and SyncBirthdayCalendar) of callForAllUsers in the dav app, but they seem to be related to calendar and contacts. We might add other callFor???User methods.

Backport to 9 strongly requested :exclamation: :grey_exclamation: !!!111einself

butonic committed:
introduce callForSeenUsers and countSeenUsers

PVince81 commented on lib/private/Repair/RepairUnmergedShares.php:
```diff
@@ -363,10 +347,9 @@ public function run(IOutput $output) {
+			$output->startProgress($this->userManager->countSeenUsers());
 
-			$this->userManager->callForAllUsers($function);
+			$this->userManager->callForSeenUsers($function);
```
won't work here because even users who haven't logged in might have broken "files_target" and duplicate shares due to the fact that some sharing API calls would set up their FS...

PVince81 commented on lib/private/User/Manager.php:
```diff
@@ -359,6 +359,88 @@ public function callForAllUsers(\Closure $callback, $search = '') {
+	 * @param string $search
+	 * @since 9.0.0
+	 */
+	public function callForSeenUsers (\Closure $callback) {
```
please add a unit test for this. It will further confirm that it works with all DBs.

PVince81 commented:
Good idea !

PVince81 commented on lib/public/IUserManager.php:
```diff
@@ -144,6 +144,21 @@ public function countUsers();
+	 * @return int
+	 * @since 9.2.0
+	 */
+	public function countSeenUsers();
```
interface change, can't backport directly

butonic replied:
hm, that still changes the return type of countUsers() to int if $onlySeen is true. I left the new methods in the interface ... should I remove them from IUserManager completely? I dislike munching everything into a single call. We can backport it that way, but in the future I want a clean API.

The callForSeenUsers call eg, does not support searching for users because we only have the userid in the preferences. callForAllUsers iterates over the user backends and the LDAP backend may search displayname or emaul in addition to the userid.

PVince81 replied:
So what to do with the interface breakage ?

In the backport you'd remove the methods from the interface ? Ok.
But then we also need to add checks everywhere `if (is_callable($userManager, 'callForSeenUsers'))` then call it, else fallback to the old way.

I'm fine adding these methods in 9.2.

PVince81 replied:
Actually adding methods in 9.2 would also require a new interface `IUserManager2` that extends `IUserManager`.

Unless we assume that no one ever bothered to implement their own user manager... there is currently no config for easy overriding without hacking server.php ...

@DeepDiver1975 thoughts ?

PVince81 replied:
Somewhere (maybe in another comment), @DeepDiver1975 answered that it is currently not possible anyway to override the interface, so changing it is ok in 9.2

butonic committed:
add tests

DeepDiver1975 commented:
I was thinking about this : With the user account table I want the call for uses to iterate only over known users.
Accessing not yet logged in users is a matter of search only from my pov.

butonic commented:
@DeepDiver1975 What is the status af the user account table?

For this PR the tests are passing and it is a clear improvement. @PVince81 @DeepDiver1975 can we merge so I can backport to 9.0.?

PVince81 commented on lib/private/User/Manager.php:
```diff
@@ -335,30 +341,117 @@ public function countUsers() {
+	 * @since 9.2.0
+	 */
+	public function countSeenUsers() {
+		$limit = 10000;
```
hard-coded limit ? should we have this as argument ?

PhilippSchaffrath replied:
If you pass that as an argument, you will have different hard-coded values for each use. Making it harder to test. If this should be variable, we should inject it as a configuration value with propel, but i think it shouldn't.

PVince81 commented on lib/private/User/Manager.php:
```diff
@@ -335,30 +341,117 @@ public function countUsers() {
+	 * @since 9.0.0
+	 */
+	public function callForSeenUsers (\Closure $callback) {
+		$limit = 1000;
```
hardcoded limit ? pass as argument ?

Also, in the count method you hard-coded to 10000 while here only 1000. Any reason ?

butonic replied:
gut feeling. count only iterates. call executes a callback ... who knows what is done in there and how much memory is leaked

butonic commented:
@PhilippSchaffrath could you review, pls?

PhilippSchaffrath commented on lib/private/User/Manager.php:
```diff
@@ -305,9 +305,15 @@ public function createUser($uid, $password) {
+	 *                   if $onlySeen is true only an int is returned
 	 */
-	public function countUsers() {
+	public function countUsers($onlySeen = false) {
```
The domain is a little inconsistent here, i would like `$onlySeen` to be called something like `$hasLoggedIn`.

PhilippSchaffrath commented on lib/private/User/Manager.php:
```diff
@@ -305,9 +305,15 @@ public function createUser($uid, $password) {
-	public function countUsers() {
+	public function countUsers($onlySeen = false) {
+		if ($onlySeen) {
+			return $this->countSeenUsers();
```
same here

butonic replied:
https://github.com/owncloud/windows_network_drive/pull/51 has already been merged we can change the name after getting this in.

PhilippSchaffrath commented on lib/private/User/Manager.php:
```diff
@@ -335,30 +341,117 @@ public function countUsers() {
-						continue;
+	public function callForAllUsers(\Closure $callback, $search = '', $onlySeen = false) {
+		if ($onlySeen) {
+			$this->callForSeenUsers($callback);
```
I would like to see a separation here. `callForAllUsers` should always call a callback for *all* users, not sometimes if parameter x is set to y only a subset of all users. Since `callForSeenUsers` already exists, why does it need to be cascaded in here?

butonic replied:
@PVince81 and @DeepDiver1975 wanted a param to make backporting easier since we don't backport API changes ...

PhilippSchaffrath commented on tests/lib/User/ManagerTest.php:
```diff
@@ -449,6 +450,66 @@ public function testCountUsersTwoBackends() {
+		$user4->delete();
+	}
+
+	public function testCallForSeenUsers() {
```
smart test, like it :+1:

DeepDiver1975 reviewed (commented):
> Actually adding methods in 9.2 would also require a new interface IUserManager2 that extends IUserManager.

This is only necessary for interfaces where we allow/expect alternative implementations out there.
I see no reason why an alternative implementation would make sense - furthermore there is no way to register one.

butonic committed:
oracle should support not null on clob

butonic commented:
I addressed most issues and also removed the sql hack for oracle because it was unnecessary to order the results in fhe first place.

@DeepDiver1975 @PVince81 @PhilippSchaffrath please review

PVince81 commented on lib/private/User/Manager.php:
```diff
@@ -359,6 +359,88 @@ public function callForAllUsers(\Closure $callback, $search = '') {
+	/**
+	 * @param \Closure $callback
+	 * @param string $search
+	 * @since 9.0.0
```
@butonic seems you missed this one

PVince81 commented:
Looks good to me, also tested with 200 LDAP users where only 4 logged in and saw that only 4 got processed for the avatar move.

@butonic please adjust the "since" tag from above and this is good from my POV :+1:

butonic committed:
since 9.2.0

butonic commented:
@PVince81 done

DeepDiver1975 commented:
:+1:

PVince81 commented:
stable9.1: ?
stable9: https://github.com/owncloud/core/issues/26415
